mitodl/edx-platform
# Reimplemented legacy remote gradebook grade export functionality

#### What are the relevant tickets?

Partially closes mitocw#74
Closes #15 
Closes #12 

#### What's this PR do?

- Adds legacy grade export functionality back into the instructor dashboard (was removed after the dogwood releases). Exposed via new instructor tab.
- Fixes security concern with requests to remote gradebook

#### Where should the reviewer start?

`lms/djangoapps/instructor/views/api.py` for back end, `lms/templates/instructor/instructor_dashboard_2/remote_gradebook.html` and `lms/static/js/instructor_dashboard/remote_gradebook.js` for front end.

#### How should this be manually tested?

- Make sure you have staff access for your test course
- Set the `ENABLE_INSTRUCTOR_REMOTE_GRADEBOOK_CONTROLS` feature to `true` (in `lms.env.json`)
- Create a `REMOTE_GRADEBOOK` setting in `lms.env.json`. This should be a dict with 2 values: 
    ```
    "REMOTE_GRADEBOOK": {
        "URL": "http://lmodproxyqa-mitodl.rhcloud.com/edx_grades",
        "DEFAULT_NAME": "STELLAR:/project/mitxdemosite"
    }
    ```

- Create `REMOTE_GRADEBOOK_USER` and `REMOTE_GRADEBOOK_PASSWORD` settings in `lms.auth.json`. Ask me or Peter for these values
- Navigate to the `Remote Gradebook` tab for some course (eg: `http://localhost:8000/courses/course-v1:edX+DemoX+Demo_Course/instructor#view-remote_gradebook`)
- Click the buttons and make sure they work. If you actually answer some questions for some assignment or exam in your course, you should see grading changes for your user when you click "Display grades for assignment"

#### Any background context you want to provide?

1. The bottom 3 buttons (starting with "Display grades...") are _really_ slow when running locally. That performance appears to vastly improve in the QA environment.
1. There's some code in here that may look like over-engineering right now (particularly in `remote_gradebook.js`), but it's likely that other legacy dashboard PRs will leverage that code
1. few key files from the dogwood release that contained the legacy dashboard functionality:
    - https://github.com/mitodl/edx-platform/blob/mitx/dogwood.3-3/lms/templates/courseware/legacy_instructor_dashboard.html
    - https://github.com/mitodl/edx-platform/blob/mitx/dogwood.3-3/lms/djangoapps/instructor/views/legacy.py#L211
